File: sdks/java/client-java-core/src/main/java/io/featurehub/client/ClientContext.java
package io.featurehub.client;

import io.featurehub.sse.model.StrategyAttributeCountryName;
import io.featurehub.sse.model.StrategyAttributeDeviceName;
import io.featurehub.sse.model.StrategyAttributePlatformName;

import java.util.List;
import java.util.Map;

public interface ClientContext {
  String get(String key, String defaultValue);

  interface ClientContextChanged {
    /**
     *
     * @param header a correctly formatted header that has been url escaped or null (if no properties)
     */
    void notify(String header);
  }

  interface  ClientContextListenerRemoval {
    void remove();
  }

  ClientContext userKey(String userKey);
  ClientContext sessionKey(String sessionKey);
  ClientContext country(StrategyAttributeCountryName countryName);
  ClientContext device(StrategyAttributeDeviceName deviceName);
  ClientContext platform(StrategyAttributePlatformName platformName);
  ClientContext version(String version);
  ClientContext attr(String name, String value);
  ClientContext attrs(String name, List<String> values);

  ClientContext clear();

  /**
   * Triggers the build and setting of this context.
   *
   * @return this
   */
  ClientContext build();

  Map<String, List<String>> context();
  String defaultPercentageKey();

  ClientContextListenerRemoval registerChangeListener(ClientContextChanged listener);
}

File: sdks/dart/client-dart-sdk/lib/src/client_context.dart
import 'package:featurehub_client_api/api.dart';

typedef ClientContextChangedHandler = Future<void> Function(String header);

class ClientContext {
  final _attributes = <String, List<String>>{};
  final _handlers = <ClientContextChangedHandler>[];

  ClientContext userKey(String userkey) {
    _attributes['userkey'] = [userkey];
    return this;
  }

  ClientContext sessionKey(String sessionKey) {
    _attributes['session'] = [sessionKey];
    return this;
  }

  ClientContext country(StrategyAttributeCountryName countryName) {
    _attributes['country'] = [countryName.name];
    return this;
  }

  ClientContext device(StrategyAttributeDeviceName device) {
    _attributes['device'] = [device.name];
    return this;
  }

  ClientContext platform(StrategyAttributePlatformName platform) {
    _attributes['platform'] = [platform.name];
    return this;
  }

  ClientContext version(String version) {
    _attributes['version'] = [version];
    return this;
  }

  ClientContext attr(String key, String value) {
    _attributes[key] = [value];
    return this;
  }

  ClientContext attrs(key, List<String> values) {
    _attributes[key] = values;
    return this;
  }

  ClientContext clear() {
    _attributes.clear();
    return this;
  }

  void build() {
    final header = generateHeader();
    for (var handler in _handlers) {
      handler(header);
    }
  }

  String generateHeader() {
    if (_attributes.isEmpty) {
      return null;
    }

    var params = _attributes.entries.map((entry) {
      return entry.key + '=' + Uri.encodeQueryComponent(entry.value.join(','));
    }).toList();
    params.sort(); // i so hate the sort function
    return params.join(',');
  }

  Future<Function> registerChangeHandler(
      ClientContextChangedHandler handler) async {
    _handlers.add(
        handler); // have to do this first in case other code triggers before this callback

    try {
      await handler(generateHeader());
      return () => {_handlers.remove(handler)};
    } catch (e) {
      _handlers.remove(handler);
      return () => {};
    }
  }
}
